src/Kestrel.Core/Internal/Http2/Http2Stream.FeatureCollection.cs

```diff
@@ -1,12 +1,34 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
 using Microsoft.AspNetCore.Server.Kestrel.Core.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
 {
-    public partial class Http2Stream : IHttp2StreamIdFeature
+    public partial class Http2Stream : IHttp2StreamIdFeature, IHttpResponseTrailersFeature
     {
+        internal HttpResponseTrailers Trailers { get; set; }
+        private IHeaderDictionary _userTrailers;
+
+        IHeaderDictionary IHttpResponseTrailersFeature.Trailers
+        {
+            get
+            {
+                if (Trailers == null)
```

Comment:
I suppose this is dependent on how _userTrailer works. But suppose the setter was called first and a non-null value was provided, this will still create a HttpResponseTrailer that will never be used.

Reply:
This has caused me to notice how incredibly inconsistent we are about using the internal or user-replaceable request and response header dictionaries for internal logic.

#2984